programs/clearing_house/src/math/lp.rs

```diff
@@ -0,0 +1,445 @@
+    } else {
+        amm.quote_asset_reserve
+            .checked_sub(new_quote_asset_reserve)
+            .ok_or_else(math_error!())?
+    };
+
+    Ok(quote_asset_reserve_output)
+}
+
+pub fn get_lp_market_position_margin(
+    lp_position: &MarketPosition,
+    amm: &AMM,
+) -> ClearingHouseResult<(MarketPosition, u128)> {
+    let total_lp_tokens = amm.sqrt_k;
+    let lp_tokens = lp_position.lp_tokens;
+
+    let mut market_position = *lp_position; // clone bc its only temporary
+    let lp_metrics = get_lp_metrics(&market_position, lp_tokens, amm)?;
+
+    // update pnl payments
+    market_position.unsettled_pnl = lp_position
+        .unsettled_pnl
+        .checked_add(lp_metrics.fee_payment)
+        .ok_or_else(math_error!())?
+        .checked_add(lp_metrics.funding_payment)
+        .ok_or_else(math_error!())?;
+
+    // worse case market position
+    // max ask: (sqrtk*1.4142 - base asset reserves) * lp share
+    // max bid: (base asset reserves - sqrtk/1.4142) * lp share
+
+    // TODO: is there a cleaner way to do this? -- maybe make it a constant?
+    let percision: f64 = 10_000_000_000_000.0; // amm percision as float
+    let sqrt_2 = (2_f64.sqrt() * percision).round() as u128;
+
+    // worse case if all asks are filled
+    let ask_bounded_k = amm
+        .sqrt_k
+        .checked_mul(sqrt_2)
+        .ok_or_else(math_error!())?
+        .checked_div(AMM_RESERVE_PRECISION)
+        .ok_or_else(math_error!())?;
+
+    let max_asks_fill = ask_bounded_k
+        .checked_sub(amm.base_asset_reserve)
+        .ok_or_else(math_error!())?;
+
+    // worse case if all bids are filled (lp is now long)
+    let bids_bounded_k = amm
+        .sqrt_k
+        .checked_mul(AMM_RESERVE_PRECISION)
+        .ok_or_else(math_error!())?
+        .checked_div(sqrt_2)
+        .ok_or_else(math_error!())?;
+
+    let max_bids_fill = amm
+        .base_asset_reserve
+        .checked_sub(bids_bounded_k)
+        .ok_or_else(math_error!())?;
+
+    // both will always be positive so its ok to compare directly
+    let net_base_asset_amount = if max_bids_fill > max_asks_fill {
+        cast_to_i128(max_bids_fill)? // lp goes long
+    } else {
+        cast_to_i128(max_asks_fill)?
+            .checked_mul(-1)
+            .ok_or_else(math_error!())? // lp is short (baa = negative)
+    };
+
+    let quote_asset_reserve_amount =
+        calculate_swap_quote_reserve_delta(amm, net_base_asset_amount)?;
+
+    let quote_asset_amount = reserve_to_asset_amount(
+        get_proportion_u128(quote_asset_reserve_amount, lp_tokens, total_lp_tokens)?,
+        amm.peg_multiplier,
+    )?;
+
+    let base_asset_amount = get_proportion_i128(net_base_asset_amount, lp_tokens, total_lp_tokens)?;
+
+    market_position.base_asset_amount = base_asset_amount;
+    market_position.quote_asset_amount = quote_asset_amount;
+
+    // additional lp margin requirements for holding lp tokens
+    let lp_margin_requirement = max(
```

Comment:
by transforming the lp position into a normal market position (base_asset_amount, quote_asset_amount, open_asks, open_bids), the margin system will be able to value it, so you don't need to do this